packages/devtools-connect/src/connect.ts

```diff
@@ -386,82 +419,163 @@ export async function connectMongoClient(
 
-  if (
-    mongoClientOptions.autoEncryption !== undefined &&
-    !mongoClientOptions.autoEncryption.bypassAutoEncryption &&
-    !mongoClientOptions.autoEncryption.bypassQueryAnalysis
-  ) {
-    // connect first without autoEncryption and serverApi options.
-    const optionsWithoutFLE = { ...mongoClientOptions };
-    delete optionsWithoutFLE.autoEncryption;
-    delete optionsWithoutFLE.serverApi;
-    const client = new MongoClientClass(uri, optionsWithoutFLE);
-    closeMongoClientWhenAuthFails(state, client);
-    await connectWithFailFast(uri, client, logger);
-    const buildInfo = await client
-      .db('admin')
-      .admin()
-      .command({ buildInfo: 1 });
-    await client.close();
     if (
-      !buildInfo.modules?.includes('enterprise') &&
-      !buildInfo.gitVersion?.match(/enterprise/)
+      mongoClientOptions.autoEncryption !== undefined &&
+      !mongoClientOptions.autoEncryption.bypassAutoEncryption &&
+      !mongoClientOptions.autoEncryption.bypassQueryAnalysis
     ) {
-      throw new MongoAutoencryptionUnavailable();
+      // connect first without autoEncryption and serverApi options.
+      const optionsWithoutFLE = { ...mongoClientOptions };
+      delete optionsWithoutFLE.autoEncryption;
+      delete optionsWithoutFLE.serverApi;
+      const client = new MongoClientClass(uri, optionsWithoutFLE);
+      closeMongoClientWhenAuthFails(state, client);
+      await connectWithFailFast(uri, client, logger);
+      const buildInfo = await client
+        .db('admin')
+        .admin()
+        .command({ buildInfo: 1 });
+      await client.close();
+      if (
+        !buildInfo.modules?.includes('enterprise') &&
+        !buildInfo.gitVersion?.match(/enterprise/)
+      ) {
+        throw new MongoAutoencryptionUnavailable();
+      }
+    }
+    uri = await resolveMongodbSrv(uri, logger);
+    const client = new MongoClientClass(uri, mongoClientOptions);
+    client.once('close', runClose);
+    closeMongoClientWhenAuthFails(state, client);
+    await connectWithFailFast(uri, client, logger);
+    if ((client as any).autoEncrypter) {
+      // Enable Devtools-specific CSFLE result decoration.
+      (client as any).autoEncrypter[
+        Symbol.for('@@mdb.decorateDecryptionResult')
+      ] = true;
     }
+    return { client, state };
+  } catch (err: unknown) {
+    await runClose();
+    throw err;
   }
-  uri = await resolveMongodbSrv(uri, logger);
-  const client = new MongoClientClass(uri, mongoClientOptions);
-  closeMongoClientWhenAuthFails(state, client);
-  await connectWithFailFast(uri, client, logger);
-  if ((client as any).autoEncrypter) {
-    // Enable Devtools-specific CSFLE result decoration.
-    (client as any).autoEncrypter[
-      Symbol.for('@@mdb.decorateDecryptionResult')
-    ] = true;
+}
+
+function hasProxyHostOption(
+  uri: string,
+  clientOptions: MongoClientOptions
+): boolean {
+  if (clientOptions.proxyHost || clientOptions.proxyPort) return true;
```

Comment:
Maybe we could rename this function to `hasProxyHostOrPortOption` option? In which cases could it be that proxyPort is enough without a host? Will be a parent host used then?

Reply:
`proxyHost` is the authoritative option, the driver forbids using `proxyPort` without `proxyHost`, e.g.

```sh
$ mongosh --quiet 'mongodb://localhost/?proxyPort=1234'
MongoParseError: Must specify proxyHost if other proxy options are passed
```

We could drop the `proxyPort` check here, I felt like it made sense to have it in case the driver ever changes this behavior (and defaults to e.g. `localhost`), but not feeling strongly about that at all